### PR TITLE
Add support for filter block

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/djc/askama/issues/955.

This PR still lacks documentation and some more tests would be nice too (if you have ideas?). But I think it's already good enough to go through a first round of reviews.

### CODE REVIEW

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:841]: Let's call this `__filter_block{id}`?

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1043]: Why?

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1150]: This should be a separate commit.

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1077]: This should go below `buf_writable()`, not above.

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1084]: If this is going to be in a separate function, let's do early return for the `Occupied` case so we can dedent the rest of it?

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1104]: This is not a great method name.

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1180]: Should be above `buf_writable()`, not below.

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:2040]: Could we avoid the caller having to wrap their `name` in `Cow`? Same for `resolver_or_self()`.

File: askama_parser/src/expr.rs
Comment (djc) [inline askama_parser/src/expr.rs:76]: That seems ugly. Why is it necessary?

File: askama_parser/src/node.rs
Comment (GuillaumeGomez) [reply askama_parser/src/node.rs:566]: Let's go for `filter_name`. :)

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1043]: Ah right, I needed to add the import to the other branch too and then my OCD triggered: it's horrible to have pattern matching with just idents and no path as then my brain doesn't know whether it means "bind everything" or "yep, look at the next one, it's actually binding each variant". I can revert that if you prefer.

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1084]: I'm not sure to see what you have in mind. Can you show me please?

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1104]: Agreed. I'm very bad at naming so any suggestion is very welcome. ^^'

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:2040]: Sure.

File: askama_parser/src/expr.rs
Comment (GuillaumeGomez) [reply askama_parser/src/expr.rs:76]: I need a variant with these two constraints:

 1. Doesn't take a reference (since it's a local variable).
 2. Doesn't make any transformation to it.

**Comment by GuillaumeGomez** [comment]
Applied all suggestions (still need a new name for the `buf_writable` method).

File: askama_derive/src/generator.rs
Comment (djc) [reply askama_derive/src/generator.rs:1043]: I can live with it in this case since it's just two arms.

File: askama_parser/src/expr.rs
Comment (djc) [reply askama_parser/src/expr.rs:76]: I mean, why do we need store any generated stuff back into an `Expr`?

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs]: This probably belongs in the other commit?

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1157]: Can we deduplicate more between the `Expr` and `Generated` arms?

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1103]: As for the name -- can you describe what this does now?

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:28]: Done.

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1083]: And of course it's obvious when seen. ^^'

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1103]: It generates the literal string and the format arguments without adding the "wrapper" (ie `::std::write!(` or `format!(`).

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1146]: Done.

File: askama_parser/src/expr.rs
Comment (GuillaumeGomez) [reply askama_parser/src/expr.rs:76]: Because we want to add the variable that was created to be used in the next `writer` use. I want this to be done in two steps: first we generate a `String` with the content of the filter and store it into a variable, then we use this variable alongside the rest of the template.

**Comment by GuillaumeGomez** [comment]
Rebasing is now too difficult so I'll rework git history once PR is approved on your side. ^^'

File: askama_derive/src/generator.rs
Comment (djc) [reply askama_derive/src/generator.rs:1097]: Maybe `prepare_format()` or some such?

**Comment by djc** [comment]
> Rebasing is now too difficult so I'll rework git history once PR is approved on your side. ^^'

The actual purpose of clean history is making review easier, so that creates a circular dependency. 😛

**Comment by GuillaumeGomez** [comment]
Oh boy. XD

Ok, let me rework the history now then. ^^'

**Comment by GuillaumeGomez** [comment]
New clean history and all comments fixed/implemented.

**Comment by GuillaumeGomez** [comment]
Ping. :)

**Comment by GuillaumeGomez** [comment]
Any news about this? :)

**Comment by GuillaumeGomez** [comment]
Ping @djc. :)

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1162]: What is a prefix commit? You mean moving this function into its own commit before the rest?

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:18]: Let's go with a struct then.

File: askama_derive/src/generator.rs
Comment (djc) [reply askama_derive/src/generator.rs:1162]: Yup!

Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1162]: Ok doing that.

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1097]: :+1:

File: askama_derive/src/generator.rs
Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1195]: That is actually much more tricky than I expected. I'd need to spend quite some time to make it work. Is it mandatory? ^^'

**Comment by GuillaumeGomez** [comment]
Updated without moving the `generate_expr_with_id` for now.

File: askama_derive/src/generator.rs
Comment (djc) [reply askama_derive/src/generator.rs:1195]: Why is it hard?

Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1195]: Mix of old and new code. I'm gonna give it another try, I'm sure I missed something obvious.

Comment (djc) [reply askama_derive/src/generator.rs:1195]: I was just looking for something like this:

https://github.com/djc/askama/compare/wip-named-expression?expand=1

Seems straightforward to me?

Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs:1195]: I was making the opposite changes first from the current PR and then tried to cherry-pick from there. It ended in a disaster. It was as terrible as it sounds. ^^'

Comment (djc) [reply askama_derive/src/generator.rs:1195]: So don't do that. 😄

Comment (GuillaumeGomez) [reply askama_derive/src/generator.rs]: Yes sir! o7

Joke aside, don't know why I did it this way... Oh well, end of the week. ^^'

**Comment by GuillaumeGomez** [comment]
Move the change in a commit on its own (and as first commit).

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:18]: Nit: as mentioned, should be defined near the bottom.

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:33]: Reading this docstring I definitely think we should just make the inner tuple a `struct` with named fields, too.

File: askama_derive/src/generator.rs
Comment (djc) [inline askama_derive/src/generator.rs:1140]: Nit: inline this binding (into the `WriteParts` initializer).

**Review by djc — COMMENTED** [review]
I think we're on the home stretch here...

**Comment by GuillaumeGomez** [comment]
Ah the push was finally taken into account by github. \o/

Now just have to wait for CI. ^^'

**Review by djc — APPROVED** [review]
Feel free to merge after the rebase -- and thanks for sticking with it.

**Comment by GuillaumeGomez** [comment]
:tada:

**Pull request merged**
